File: Answer/3531.统计被覆盖的建筑.go
/*
 * @lc app=leetcode.cn id=3531 lang=golang
 *
 * [3531] 统计被覆盖的建筑
 */

// @lc code=start
func countCoveredBuildings(n int, buildings [][]int) int {
	//Key：横坐标；Value：纵坐标范围
	ver := make(map[int][2]int)
	//Key：纵坐标；Value：横坐标范围
	hor := make(map[int][2]int)
	for _, building := range buildings {
		x := building[0]
		y := building[1]
		if existing, exists := ver[x]; exists {
			ver[x] = [2]int{int(math.Min(float64(existing[0]), float64(y))), int(math.Max(float64(existing[1]), float64(y)))}
		} else {
			ver[x] = [2]int{y, y}
		}

		if existing, exists := hor[y]; exists {
			hor[y] = [2]int{int(math.Min(float64(existing[0]), float64(x))), int(math.Max(float64(existing[1]), float64(x)))}
		} else {
			hor[y] = [2]int{x, x}
		}
	}
	res := 0
	for _, building := range buildings {
		x := building[0]
		y := building[1]
		if ver[x][0] < y && ver[x][1] > y && hor[y][0] < x && hor[y][1] > x {
			res += 1
		}
	}
	return res
}

// @lc code=end

File: Answer/3531.统计被覆盖的建筑.js
/*
 * @lc app=leetcode.cn id=3531 lang=javascript
 *
 * [3531] 统计被覆盖的建筑
 */

// @lc code=start
/**
 * @param {number} n
 * @param {number[][]} buildings
 * @return {number}
 */
var countCoveredBuildings = function (n, buildings) {
    //Key：横坐标；Value：纵坐标范围
    let ver = new Map();
    //Key：纵坐标；Value：横坐标范围
    let hor = new Map();
    buildings.forEach(building => {
        let x = building[0];
        let y = building[1];
        if (ver.has(x)) {
            ver.get(x)[0] = Math.min(ver.get(x)[0], y);
            ver.get(x)[1] = Math.max(ver.get(x)[1], y);
        } else {
            ver.set(x, [y, y]);
        }
        if (hor.has(y)) {
            hor.get(y)[0] = Math.min(hor.get(y)[0], x);
            hor.get(y)[1] = Math.max(hor.get(y)[1], x);
        } else {
            hor.set(y, [x, x]);
        }
    });
    let res = 0;
    buildings.forEach(building => {
        var x = building[0];
        var y = building[1];
        if (ver.get(x)[0] < y && ver.get(x)[1] > y && hor.get(y)[0] < x && hor.get(y)[1] > x) {
            res++;
        }
    });
    return res;
};
// @lc code=end

